hypothesis-python/tests/conjecture/common.py

```diff
@@ -67,3 +72,153 @@ def accept(f):
             )
 
     return accept
+
+
+def fresh_data():
+    return ConjectureData(BUFFER_SIZE, prefix=b"", random=Random())
+
+
+@st.composite
+def draw_integer_kwargs(
+    draw,
+    *,
+    use_min_value=True,
+    use_max_value=True,
+    use_shrink_towards=True,
+    use_weights=True,
+    use_forced=False,
+):
+    min_value = None
+    max_value = None
+    shrink_towards = 0
+    weights = None
+
+    # this generation is complicated to deal with maintaining any combination of
+    # the following invariants, depending on which parameters are passed:
+    #
+    # (1) min_value <= forced <= max_value
+    # (2) max_value - min_value + 1 == len(weights)
+    # (3) len(weights) <= 1024
+
+    forced = draw(st.integers()) if use_forced else None
+    if use_weights:
+        assert use_max_value
+        assert use_min_value
+        # handle the weights case entirely independently from the non-weights case.
+        # We'll treat the weights as our "key" draw and base all other draws on that.
+
+        # weights doesn't play well with super small floats, so exclude <.01
+        weights = draw(st.lists(st.floats(0.01, 1), min_size=1, max_size=1024))
+
+        # we additionally pick a central value (if not forced), and then the index
+        # into the weights at which it can be found - aka the min-value offset.
+        center = forced if use_forced else draw(st.integers())
+        min_value = center - draw(st.integers(0, len(weights) - 1))
+        max_value = min_value + len(weights) - 1
+    else:
+        if use_min_value:
+            min_value = draw(st.integers(max_value=forced))
+        if use_max_value:
+            min_vals = []
+            if min_value is not None:
+                min_vals.append(min_value)
+            if forced is not None:
+                min_vals.append(forced)
+            min_val = max(min_vals) if min_vals else None
+            max_value = draw(st.integers(min_value=min_val))
+
+    if use_shrink_towards:
+        shrink_towards = draw(st.integers())
+
+    if forced is not None:
+        assume((forced - shrink_towards).bit_length() < 128)
+
+    return {
+        "min_value": min_value,
+        "max_value": max_value,
+        "shrink_towards": shrink_towards,
+        "weights": weights,
+        "forced": forced,
+    }
+
+
+@st.composite
+def draw_string_kwargs(draw, *, use_min_size=True, use_max_size=True, use_forced=False):
+    interval_set = draw(intervals())
+    # TODO relax this restriction once we handle empty pseudo-choices in the ir
+    assume(len(interval_set) > 0)
```

Comment:
it's a bit annoying to address this in this pull, so I'd prefer to wait for #3818 .